### PR TITLE
Native verification

### PR DESCRIPTION
**Pull request by dark64**
- Added verify command to the CLI
- Implemented native verification for groth16 (utilizing bellman)
- Simplified parsing of verification key by adding a key-value parser.
- Added native verification for libsnark schemes (gm17 and pghr13)

**Note**: This PR had to change verification key by adding an extra key called `vk.raw` that contains raw hex encoded bytes of underlying backend-specific verification key. Same logic applies for the proof.

### CODE REVIEW

File: zokrates_core/src/proof_system/bn128/utils/parser.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+use std::collections::HashMap;
+
+pub struct KeyValueParser;
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/bn128/utils/parser.rs]: I'm wondering if we can get this done with the csv crate (we already use to (de)serialize the witness) by setting the delimiter here https://docs.rs/csv/1.1.3/csv/struct.ReaderBuilder.html#method.delimiter

File: zokrates_core/src/proof_system/bn128/g16.rs [hunk 2]
```diff
@@ -222,35 +220,47 @@ impl ProofSystem for G16 {
...
-            .collect::<Vec<_>>()
-            .join("\n")
-    );
+fn serialize_vk(vk: VerifyingKey<Bn256>, include_raw: bool) -> String {
+    let mut writer = csv::WriterBuilder::new()
+        .delimiter(b'=')
+        .from_writer(vec![]);
+
+    writer
+        .write_record(vec!["vk.alpha", parse_g1_hex(&vk.alpha_g1).as_str()])
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/bn128/g16.rs]: detail, but is there a way out of creating a vec here?

Comment (dark64) [reply zokrates_core/src/proof_system/bn128/g16.rs]: yes we can use array here

File: zokrates_core/lib/gm17.cpp [hunk 3]
```diff
@@ -8,6 +8,7 @@
 #include "gm17.hpp"
 #include "util.hpp"
 #include <cassert>
+#include <fstream>
```
Comment (Schaeff) [inline zokrates_core/lib/gm17.cpp:11]: is this needed? we're trying to keep the backend from requiring a fs

File: zokrates_core/src/proof_system/bn128/gm17.rs [hunk 4]
```diff
@@ -16,26 +21,46 @@ impl GM17 {
...
+#[derive(Serialize, Deserialize)]
+struct GM17ProofPoints {
+    a: G1PairingPoint,
+    b: G2PairingPoint,
+    c: G1PairingPoint,
+}
+
+impl Proof<GM17ProofPoints> {
+    fn from_json(str: &str) -> Self {
+        serde_json::from_str(str).unwrap()
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/bn128/gm17.rs:33]: Just making sure we're not unwrapping on a user error here?

File: zokrates_core/src/proof_system/bn128/pghr13.rs [hunk 5]
```diff
@@ -16,26 +20,51 @@ impl PGHR13 {
...
+    b_p: G1PairingPoint,
+    c: G1PairingPoint,
+    c_p: G1PairingPoint,
+    h: G1PairingPoint,
+    k: G1PairingPoint,
+}
+
+impl Proof<PGHR13ProofPoints> {
+    fn from_json(str: &str) -> Self {
+        serde_json::from_str(str).unwrap()
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/bn128/pghr13.rs:37]: same here

File: zokrates_core/src/proof_system/bn128/utils/parser.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
+use std::collections::HashMap;
+
+pub fn parse_vk(vk: String) -> HashMap<String, String> {
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/bn128/utils/parser.rs:3]: Same here, shouldn't this return a `Result` ?

File: zokrates_core/src/proof_system/bn128/utils/parser.rs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        }
+    }
+    map
+}
+
+#[cfg(test)]
+mod tests {
+    use proof_system::bn128::utils::parser::parse_vk;
+
+    #[test]
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/bn128/utils/parser.rs:26]: this needs more negative tests with invalid files

File: zokrates_core/src/proof_system/mod.rs [hunk 8]
```diff
@@ -34,5 +34,7 @@ pub trait ProofSystem {
         proving_key: Vec<u8>,
     ) -> String;
 
-    fn export_solidity_verifier(&self, vk: String, is_abiv2: bool) -> String;
+    fn export_solidity_verifier(&self, vk: String, abi_v2: bool) -> String;
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/mod.rs:37]: I think `is_` is a good prefix for boolean variables. The Rust way would actually be to define 
```
enum Abi {
   V1,
   V2
}
```

Maybe we should consider that?

File: zokrates_core/src/proof_system/bn128/g16.rs [hunk 9]
```diff
@@ -20,25 +29,42 @@ impl G16 {
...
+}
+
+impl G16ProofPoints {
+    fn new(a: G1PairingPoint, b: G2PairingPoint, c: G1PairingPoint) -> Self {
+        G16ProofPoints { a, b, c }
+    }
+}
+
+impl Proof<G16ProofPoints> {
+    fn from_json(str: &str) -> Self {
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/bn128/g16.rs:46]: same here

File: zokrates_core/src/proof_system/bn128/utils/ffi.rs [hunk 10]
```diff
@@ -20,14 +20,27 @@ extern "C" {
 }
 
 impl Buffer {
-    pub fn from_vec(v: &mut Vec<u8>) -> Buffer {
-        let length = v.len() as i32;
+    pub unsafe fn from_vec(v: &Vec<u8>) -> Buffer {
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/bn128/utils/ffi.rs:23]: can we make this into an unsafe block of minimal size?

File: zokrates_core/src/proof_system/bn128/utils/libsnark.rs [hunk 11]
```diff
@@ -14,6 +14,19 @@ fn vec_as_u8_32_array(vec: &Vec<u8>) -> [u8; 32] {
     array
 }
 
+pub fn prepare_public_inputs(public_inputs: Vec<&str>) -> (Vec<[u8; 32]>, usize) {
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/bn128/utils/libsnark.rs:17]: here also I'm concerned about user errors leading to panics

File: zokrates_cli/src/bin.rs [hunk 12]
```diff
@@ -263,6 +263,34 @@ fn cli() -> Result<(), String> {
             .required(true)
         )
     )
+    .subcommand(SubCommand::with_name("verify")
+        .about("Verifies a given proof with the given constraint system and verification key")
```
[the remark this answers is not in the document]
Comment (dark64) [reply zokrates_cli/src/bin.rs]: Fixed

File: .circleci/config.yml [hunk 13]
```diff
@@ -42,9 +42,9 @@ jobs:
       - restore_cache:
           keys:
             - v4-cargo-cache-{{ arch }}-{{ checksum "Cargo.lock" }}
-      - run:
-          name: Check format
-          command: rustup component add rustfmt; cargo fmt --all -- --check
+      # - run:
```
[the remark this answers is not in the document]
Comment (dark64) [reply .circleci/config.yml]: Done

File: zokrates_core/lib/gm17.cpp (hunk 3 above)
Comment (dark64) [reply zokrates_core/lib/gm17.cpp]: Good catch

File: zokrates_core/src/proof_system/bn128/utils/parser.rs (hunk 7 above)
Comment (dark64) [reply zokrates_core/src/proof_system/bn128/utils/parser.rs:25]: Added 2 more tests

File: zokrates_core/src/proof_system/bn128/pghr13.rs (hunk 5 above)
Comment (dark64) [reply zokrates_core/src/proof_system/bn128/pghr13.rs]: Moved this method to https://github.com/Zokrates/ZoKrates/pull/583/files#diff-189d3ba6c93acd77029cea5c8a6a386aR26

File: zokrates_core/src/proof_system/bn128/gm17.rs (hunk 4 above)
Comment (dark64) [reply zokrates_core/src/proof_system/bn128/gm17.rs]: Moved this method to https://github.com/Zokrates/ZoKrates/pull/583/files#diff-189d3ba6c93acd77029cea5c8a6a386aR26

File: zokrates_core/src/proof_system/mod.rs (hunk 8 above)
Comment (dark64) [reply zokrates_core/src/proof_system/mod.rs]: Introducing enum is the way to go here imo

File: zokrates_core/src/proof_system/bn128/g16.rs (hunk 9 above)
Comment (dark64) [reply zokrates_core/src/proof_system/bn128/g16.rs]: Moved this method to https://github.com/Zokrates/ZoKrates/pull/583/files#diff-189d3ba6c93acd77029cea5c8a6a386aR26

File: zokrates_core/src/proof_system/bn128/utils/ffi.rs (hunk 10 above)
Comment (dark64) [reply zokrates_core/src/proof_system/bn128/utils/ffi.rs]: This method does not need to be unsafe, it does use unsafe methods under the hood though. Removed the unsafe keyword for clarity.

File: zokrates_core/src/proof_system/bn128/utils/libsnark.rs (hunk 11 above)
Comment (dark64) [reply zokrates_core/src/proof_system/bn128/utils/libsnark.rs]: This method now accepts already parsed fields https://github.com/Zokrates/ZoKrates/pull/583/files#diff-26394101635fe9e8e900553bde8952baR17

File: zokrates_core/src/proof_system/bn128/g16.rs [hunk 14]
```diff
@@ -18,6 +18,7 @@ use crate::proof_system::bn128::utils::solidity::{
 };
 use crate::proof_system::{ProofSystem, SetupKeypair};
 use proof_system::bn128::{G1PairingPoint, G2PairingPoint, Proof};
+use proof_system::AbiVersion;
```
Comment (Schaeff) [inline zokrates_core/src/proof_system/bn128/g16.rs:21]: let's call this `SolidityAbi(Version)?`

File: zokrates_js/index.d.ts [hunk 15]
```diff
@@ -20,13 +20,14 @@ declare module 'zokrates-js' {
     pk: Uint8Array,
   }
 
+  export type AbiVersion = "v1" | "v2";
```
Comment (Schaeff) [inline zokrates_js/index.d.ts:23]: Nice!

**Pull request merged**